osh/cmd_eval.py

```diff
@@ -1896,6 +1901,13 @@ def _Execute(self, node):
                 self.errfmt.PrettyPrintError(e, prefix='failglob: ')
                 status = 1  # another redirect word eval error
                 cmd_st.check_errexit = True  # failglob + errexit
+            except error.VarSubFailure as e:  # e.g. echo hi > ${!undef}
+                if not e.HasLocation():  # Last resort!
+                    e.location = self.mem.GetFallbackLocation()
+                self.errfmt.PrettyPrintError(e)
+                status = 1  # another redirect word eval error
+                cmd_st.check_errexit = True  # errexit [TODO: what is this??]
+
```

Comment:
I think if failglob does it, it's probably right

I am pretty sure if you don't set it, it will be easy to see it in the tests.  If not, we may need to rethink that a bit

---

We could also generalize this to "WordFailure" -- i.e. both FailGlob and VarSub --   I am not sure if there are other cases where it comes up --   I think there might be 1 or 2 more

Reply:
Sorry for the delay. I confirmed that this `cmd_st.check_errexit = True` is responsible for `errexit` of e.g. `a=${!undef}` and `for a in ${!undef}`. I fixed the comment: [`9386e20..9742e3e`](https://github.com/oils-for-unix/oils/compare/9386e203706ec600f844279cf641ffea0e98650d..9742e3ee7aa5fe8b35168af5238b86cf8fb362be).

----

> We could also generalize this to "WordFailure" -- i.e. both FailGlob and VarSub -- I am not sure if there are other cases where it comes up -- I think there might be 1 or 2 more

However, the error messages shown for FailGlob and VarSubFailure are different. We have `prefix='failglob: '` for `FailGlob`:

https://github.com/oils-for-unix/oils/blob/55a49cad13df1b14391708b6700244e49b3d1a3d/osh/cmd_eval.py#L1538

In commit 8894c8c6a3dfdbee729e2a0c5932e0ce80918e4c, I decided to subclass `FailGlob` and `VarSubFailure` as derived classes of `WordFailure` and included `'failglob: '` in the message of `FailGlob`. Does this work? If not, I'll drop this commit.